crates/notion-core/src/tool/binary.rs

```diff
@@ -1,83 +1,70 @@
-
-        // first try to use the project toolchain
-        if let Some(project) = session.project()? {
-            // check if the executable is a direct dependency
-            if project.has_direct_bin(&params.executable)? {
-                // use the full path to the file
-                let mut path_to_bin = project.local_bin_dir();
-                path_to_bin.push(&params.executable);
-
-                // if we're in a pinned project, use the project's platform.
-                if let Some(ref platform) = project.platform() {
-                    let image = platform.checkout(session)?;
-                    return Ok(Self::from_components(
-                        &path_to_bin.as_os_str(),
-                        params.args,
-                        &image.path()?,
-                    ));
-                }
+    // first try to use the project toolchain
+    if let Some(project) = session.project()? {
+        // check if the executable is a direct dependency
+        if project.has_direct_bin(&exe)? {
+            // use the full path to the file
+            let mut path_to_bin = project.local_bin_dir();
+            path_to_bin.push(&exe);
 
-                // otherwise use the user platform.
-                if let Some(ref platform) = session.user_platform()? {
-                    let image = platform.checkout(session)?;
-                    return Ok(Self::from_components(
-                        &path_to_bin.as_os_str(),
-                        params.args,
-                        &image.path()?,
-                    ));
-                }
+            // if we're in a pinned project, use the project's platform.
+            if let Some(ref platform) = project.platform() {
+                let image = platform.checkout(session)?;
+                return Ok(ToolCommand::direct(
+                    &path_to_bin.as_os_str(),
+                    args,
+                    &image.path()?,
+                ));
+            }
 
-                // if there's no user platform selected, fail.
-                throw!(ErrorDetails::NoPlatform);
+            // otherwise use the user platform.
+            if let Some(ref platform) = session.user_platform()? {
+                let image = platform.checkout(session)?;
+                return Ok(ToolCommand::direct(
+                    &path_to_bin.as_os_str(),
+                    args,
+                    &image.path()?,
+                ));
             }
-        }
 
-        // try to use the user toolchain
-        if let Some(user_tool) = session.get_user_tool(&params.executable)? {
-            return Ok(Self::from_components(
-                &user_tool.bin_path.as_os_str(),
-                params.args,
-                &user_tool.image.path()?,
-            ));
+            // if there's no user platform selected, pass through to existing PATH.
+            return ToolCommand::passthrough(&exe, args, ErrorDetails::NoPlatform);
```

Comment:
I think this return should be removed, so the code falls through to looking for a user tool (instead of trying to pass through to the system at this point).

In the following scenario:
* the user installs Jetson
* the user runs `notion install tsc`, for instance

At that point, `tsc` is installed as a user tool, but there is no default version of Node set for the user (because package install doesn't set the default, it just makes sure there is a platform downloaded).

Then when the user runs `tsc <whatever>` in the project, it will fail, because it doesn't get far enough to look for the user tool.

Reply:
My only concern is that we're then executing the user version of a tool inside of a package that we know for a fact has that tool as a dependency, because we don't have a way to invoke that local version. Of note, the current change matches the previous behavior, where we would throw an error here instead of falling through to checking for a user tool.

I'm a bit on the fence about whether making it so `jetson install typescript` lets you run `tsc` anywhere makes more sense than preventing users from accidentally running a global tool in a local workspace, which is one of the goals of Jetson as a product. That said, we are opening that protection up a little anyway by using the passthrough, so I suppose it's not unreasonable.

What about a third option: We check for the user tool settings, and if we find them, we execute the _local_ version (using `path_to_bin`) with the platform image provided by the `bin_config`. That matches the approach we would take if we found the user platform but not a local platform, just lining it up with the tool version. In fact, that might actually be better as the 2nd check, instead of the 3rd.

Reply:
That might actually be a better option.

After thinking more about what I was saying above, I'm not sure it's a good idea to run the user tool version in this case, because it somewhat breaks the Jetson model of how things should work.

Reply:
I like @charlespierce's angle here, but I'd suggest *warning* in that case. It makes things *work* so users' flow isn't disrupted, but lets them know that there's something unusual in the flow.

Reply:
I'm a little wary of warnings in places that will likely be called often, as that tends to result in 1 of 2 reactions:

1. The user sees the warning, thinks everything is broken, and files an issue or stops using Jetson.
2. The user realizes the warning isn't relevant and mentally tunes it out, making future warnings less impactful.

Though thinking on it a bit more, I'm actually not sure that looking for a user tool is the right approach at all. If we are in a project that has a local version of a tool, we should definitely be calling that local version. In order to minimize the chance of weird node version mismatch bugs, we should try to use the same version of Node with which the local version was installed:

- If the project doesn't have a `toolchain` key, then it most likely wasn't installed using a project-local Node version.
- If the user doesn't have a default Node version, then it wasn't installed using the user default Node version.
- The only thing left is that it was installed by `npm` or `yarn` that already existed on the system, either before Jetson was installed or through the passthrough mechanism.

In the last case, we should try to use the same Node version that already existed by passing through as well. We should, however, still be executing the local tool version, so we should update this passthrough to use the full path to the binary, instead of just the tool name (`bin_path` instead of `exe` in the code).

The user tools are installed in a sandbox that isn't connected to the versions of `npm` or `yarn` that the user executes, while local tools are always installed through `npm install` or `yarn`, so I don't think we should be looking towards the sandbox when executing local tools.

Reply:
Fair point on warnings; my inclination is generally to give more rather than less information. It's *useful* to provide info and not only silently do the "right" thing where possible, because it gives the user opportunities to opt into (what we would describe as) better alternatives. However, you're quite right about the two failures modes, and designing a solution that solves all of those is non-trivial.

I'll keep mulling on the design question here, and I'm also now thinking on how we might give users one-time or silence-able warnings for scenarios like that.

Reply:
For providing information, I think this would be a great use-case for an expanded `--verbose` flag, as described in #362, if the user opts into the additional information, we can list which platform we are choosing and where it came from as part of the output.

For tools that's a little tricky, we may need an environment variable like `JETSON_VERBOSE` or something to switch that on, but we could provide a way for users to opt-in to seeing more information about the decisions Jetson is making

Reply:
Sorry I lost track of this PR. Just looked at the code again to refresh my memory on this flow. I think this echoes what @charlespierce is saying above.

At this point, the user is:
* in a project,
* running a shimmed executable,
* where that executable is a dependency of that project,
* and the project does not have `toolchain`,
* and the user does not have a default `node`

The reason there is a shim at this point is that the user ran `notion install <package>`, so they have this executable available as a user tool.

But:
* this project does not use Notion (there's no `toolchain`),
* this project's dependencies were not installed with the user version of `node`, because there is no default

So I think we should do what @charlespierce is suggesting above - try to execute the locally installed version of the executable, using the system Node, which is presumably what was used to install it.

I would prefer to give some kind of info/warning that Notion is running the local version, with a CTA of `Run 'notion install node' or 'notion pin node' to fix this warning`, something like that, while still trying to do the right thing.

Reply:
On the warning subject, what we do in this PR is keep the error we would have thrown before the passthrough (`NoPlatform` in this case), which tells the user that `node` isn't installed and calls out `notion install node` / `notion pin node`. Then if the passthrough fails for any reason, we show that error. So if the user has their system set up and hasn't opted in to using Notion fully, then it should continue to work exactly as they expect. If it doesn't, then we can guide them to how to fix it.

I'm wary of providing warnings in cases where it actually works, because then users get annoyed at seeing warnings where they don't actually affect the execution of the tool.

Reply:
@mikrostew And yes, I didn't say in the last comment, but your summary describes exactly the situation we are in, and the current solution to it.

Reply:
👍